File: test/Dapr.IntegrationTest.Workflow/SimpleWorkflowTests.cs
// ------------------------------------------------------------------------
// Copyright 2025 The Dapr Authors
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//     http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//  ------------------------------------------------------------------------

using Dapr.TestContainers.Common;
using Dapr.TestContainers.Common.Options;
using Dapr.Workflow;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;

namespace Dapr.IntegrationTest.Workflow;

public sealed class SimpleWorkflowTests
{
    [Fact]
    public async Task ShouldHandleSimpleWorkflow()
    {
        var options = new DaprRuntimeOptions();
        var componentsDir = TestDirectoryManager.CreateTestDirectory("workflow-components");
        var workflowInstanceId = Guid.NewGuid().ToString();

        var harness = new DaprHarnessBuilder(options).BuildWorkflow(componentsDir);
        await using var testApp = await DaprHarnessBuilder.ForHarness(harness)
            .ConfigureServices(builder =>
            {
                builder.Services.AddDaprWorkflowBuilder(
                    configureRuntime: opt =>
                    {
                        opt.RegisterWorkflow<TestWorkflow>();
                        opt.RegisterActivity<DoublingActivity>();
                    },
                    configureClient: (sp, clientBuilder) =>
                    {
                        var config = sp.GetRequiredService<IConfiguration>();
                        var grpcEndpoint = config["DAPR_GRPC_ENDPOINT"];
                        if (!string.IsNullOrEmpty(grpcEndpoint))
                            clientBuilder.UseGrpcEndpoint(grpcEndpoint);
                    });
            })
            .BuildAndStartAsync();

        // Clean test logic
        using var scope = testApp.CreateScope();
        var daprWorkflowClient = scope.ServiceProvider.GetRequiredService<DaprWorkflowClient>();

        // Start the workflow
        const int startingValue = 8;

        await daprWorkflowClient.ScheduleNewWorkflowAsync(nameof(TestWorkflow), workflowInstanceId, startingValue);
        var result = await daprWorkflowClient.WaitForWorkflowCompletionAsync(workflowInstanceId, true);

        Assert.Equal(WorkflowRuntimeStatus.Completed, result.RuntimeStatus);
        var resultValue = result.ReadOutputAs<int>();
        Assert.Equal(16, resultValue);
    }

    private sealed class DoublingActivity : WorkflowActivity<int, int>
    {
        public override Task<int> RunAsync(WorkflowActivityContext context, int input)
        {
            var square = input * 2;
            return Task.FromResult(square);
        }
    }

    private sealed class TestWorkflow : Workflow<int, int>
    {
        public override async Task<int> RunAsync(WorkflowContext context, int input)
        {
            var result = await context.CallActivityAsync<int>(nameof(DoublingActivity), input);
            return result;
        }
    }
}

File: test/Dapr.IntegrationTest.FSharp.Workflow/SimpleWorkflowTests.fs
#nowarn "FS3261"
namespace Dapr.IntegrationTest.FSharp.Workflow

open System
open System.Threading.Tasks
open Dapr.Testcontainers.Common
open Dapr.Testcontainers.Harnesses
open Dapr.Workflow
open Dapr.Workflow.Registration
open Microsoft.Extensions.Configuration
open Microsoft.Extensions.DependencyInjection
open Xunit

type SimpleWorkflowTests() =

    [<Fact>]
    member this.ShouldHandleSimpleWorkflow() = task {
        let componentsDir = TestDirectoryManager.CreateTestDirectory("fsharp-workflow-components")
        let workflowInstanceId = Guid.NewGuid().ToString()

        use! env = DaprTestEnvironment.CreateWithPooledNetworkAsync(
            needsActorState = true,
            cancellationToken = TestContext.Current.CancellationToken)
        do! env.StartAsync(TestContext.Current.CancellationToken)

        let harness =
            DaprHarnessBuilder(componentsDir)
                .WithEnvironment(env)
                .BuildWorkflow()

        let! testApp =
            DaprHarnessBuilder.ForHarness(harness)
                .ConfigureServices(fun builder ->
                    builder.Services.AddDaprWorkflowBuilder(
                        Action<WorkflowRuntimeOptions>(fun opt ->
                            opt.RegisterWorkflow<TestWorkflow>()
                            opt.RegisterActivity<DoublingActivity>()),
                        Action<IServiceProvider, DaprWorkflowClientBuilder>(fun sp clientBuilder ->
                            let config = sp.GetRequiredService<IConfiguration>()
                            let grpcEndpoint = config.["DAPR_GRPC_ENDPOINT"]
                            if not (String.IsNullOrEmpty(grpcEndpoint)) then
                                clientBuilder.UseGrpcEndpoint(grpcEndpoint) |> ignore)) |> ignore)
                .WithDaprStartupOrder(true)
                .BuildAndStartAsync()
        use testApp = testApp

        use scope = testApp.CreateScope()
        let daprWorkflowClient = scope.ServiceProvider.GetRequiredService<DaprWorkflowClient>()

        let startingValue = 8
        let! _ = daprWorkflowClient.ScheduleNewWorkflowAsync(typeof<TestWorkflow>.Name, workflowInstanceId, startingValue)
        let! result = daprWorkflowClient.WaitForWorkflowCompletionAsync(
            workflowInstanceId, true, TestContext.Current.CancellationToken)

        Assert.Equal(WorkflowRuntimeStatus.Completed, result.RuntimeStatus)
        let resultValue = result.ReadOutputAs<int>()
        Assert.Equal(16, resultValue)
    }

and DoublingActivity() =
    inherit WorkflowActivity<int, int>()
    override _.RunAsync(_: WorkflowActivityContext, input: int) : Task<int> =
        Task.FromResult<int>(input * 2)

and TestWorkflow() =
    inherit Workflow<int, int>()
    override _.RunAsync(context: WorkflowContext, input: int) : Task<int> =
        task {
            let! result = context.CallActivityAsync<int>(typeof<DoublingActivity>.Name, input)
            return result
        }
